Use cudnn when possible instead of conv2d

fvisin opened the pull request:
This is a merge of an old work by @ballasn.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -2987,15 +2995,27 @@ def initialize_transformer(self, rng):
+                    kernel_shape=self.kernel_shape,
+                    subsample=self.kernel_stride,
+                    border_mode=self.border_mode,
+                    rng=rng)
```
Here, we completely ignore `pad`, we should have a meaningful error message instead if `pad` is specified but `cudnn2d` is not available.

fvisin replied:
I removed the padding.

lamblin replied:
OK.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -3059,6 +3079,14 @@ def set_input_space(self, space):
+                                                     self.kernel_stride,
+                                                     self.pad)]
+
         if self.border_mode == 'valid':
```
We should make sure 'border_mode' and 'pad' are not passed at the same time, I guess.
Also, the `output_shape` will be overwritten by default.

fvisin replied:
(I got rid of pad, following Fred's suggestion, so this comment does not apply anymore to this PR. I keep it for future reference.)

`border_mode` and `pad` can be passed together, as the latter augments the data but does not guarantee to do it such as the final size is a multiple of the kernel convolution (i.e., you might still need to either pad more or ignore part of the _padded image_)

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -3186,6 +3214,25 @@ def get_weights_topo(self):
+        return OrderedDict([('kernel_norms_min',  row_norms.min()),
+                            ('kernel_norms_mean', row_norms.mean()),
+                            ('kernel_norms_max',  row_norms.max()), ])
+
```
That method had been removed, we should not put it back here.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -3204,6 +3251,14 @@ def get_layer_monitoring_channels(self, state_below=None,
                            ('kernel_norms_max', row_norms.max()),
                            ])
 
+        if cudnn2d:
```
Why is this needed?

fvisin replied:
I have to admit I have no idea. It was part of the original code so I merged it, but it doesn't seem to be of any use. I will remove it.

lamblin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,168 @@
+                 input_space,
+                 output_axes=('b', 'c', 0, 1),
+                 subsample=(1, 1),
+                 pad=(1,1),
```
Why that for default padding, and not `(0, 0)`?

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -3059,6 +3079,14 @@ def set_input_space(self, space):
 
+        if self.pad != (0, 0):
+            output_shape = \
+                [int(np.ceil((i_sh + 2. * k_pad - k_sh) / float(k_st))) + 1
```
Actually, I'm not sure the `ceil` part is right, as it does not give the same result as the integer division below when pad is `(0, 0)`.

fvisin replied:
(I got rid of pad, following Fred's suggestion, so this comment does not apply anymore to this PR. I keep it for future reference.)

If pad is (0, 0) this is not executed.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -3246,9 +3301,19 @@ def fprop(self, state_below):
+                    # TODO check if this can be ported to the GPU one
+                    # TODO check axes: should they be B01C or BC01 or will this
+                    # always work?
+                    p = downsample.max_pool_2d(d, self.pool_shape,
```
You are not passing `self.pool_stride` here, which is why the dimensions are not right.

fvisin replied:
Great catch! :) 
Actually, does it even make sense to use `downsample.max_pool_2d` instead of `max_pool` as in the "non cudnn2d" case? Apparently there is no reason. I am not sure if the data will be in bc01 format, though I think that the data has been already reshaped in that form somewhere since the "normal" convolution does not force you to provide data in bc01 format. If you don't already know the answer, I will check if this is the case.

nouiz commented:
I just want to bring to your attention that the pooling operator is better handled (and have fix when stride > window size) in this PR: https://github.com/lisa-lab/pylearn2/pull/1361

Maybe you should remove the related change to this part? I just need to check if tests I ran passed to merge it.

fvisin commented:
Thank you @lamblin and @nouiz for the super-fast support!
 
I removed the padding as I don't need it and it didn't seem to be handled very well.
Also, I removed the "special case" pooling falling back to the conv2d behavior that seems correct also in the cudnn case. This will allow to merge the other PR flawlessly. I am running an experiment on the new code, I will comment further if something goes wrong.

nouiz commented:
The other PR with change for max pooling was merged. Can you remove the change related to that that you have and rebase on master?

JesseLivezey commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,162 @@
+        op_axes = ('b', 'c', 0, 1)
+
+        if tuple(axes) != op_axes:
+            x = x.dimshuffle(
```
I think you could write this as
```
x = x.dimshuffle(*[axes.index(ax) for ax in op_axes])
```
and it would be more concise and reuse the ```op_axes``` definition.

fvisin replied:
Yes, you are right. This commit was a very preliminary release, it needs refactoring. Thank you for pointing that out :)

JesseLivezey commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,162 @@
+        assert len(axes) == 4
+
+        if tuple(axes) != op_axes:
+            rval = rval.dimshuffle(
```
Also here.

fvisin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,203 @@
+    ----------
+    filters : Theano shared variable
+        4D-tensor of shape (out channels, in channels, rows, cols)
+    batch_size : int
```
Do we need batch_size? Can we get it from the input space?

lamblin replied:
The batch size is not part of the space. For the original Conv2D, batch_size could be None, indicating that the batch size was not constant. Can you check if this is supported by cudnn as well?

fvisin replied:
Uhm....cudnn is this file, I am confused. :) 

batch_size is only used to compute the img_shape parameter passed to the _original Conv2D_ (`pylearn2.packaged_dependencies.theano_linear.conv2d`) constructor. The exact same thing is done in `pylearn2.linear.conv2d`.

lamblin replied:
I mean, by the underlying Theano Op, GpuDnnConv.

fvisin replied:
The original `Conv2D` uses `batch_size` to define `image_shape`, that is then used as an argument for the constructors of both the `theano.tensor.nnet.conv` function and the `pylearn2.packaged_dependencies.theano_linear.conv2d` function.

`GpuDnnConv` seems not to use the batch_size at all. The underlying Theano Op [[doc]](http://deeplearning.net/software/theano/library/sandbox/cuda/dnn.html) [[code]](https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/dnn.py) takes only 3 parameters: `image`, `kernel`, `descr`.

Does this answer to your question?

lamblin replied:
(That was addressed.)

fvisin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,203 @@
+    input_space : Space
+        The Space of the input data
+    output_axes : tuple, optional
+        The requested output axes. If not specified `bc01` will be used.
```
Does it make sense to default to bc01? Wouldn't it be better to default to the same axes as the input?

lamblin replied:
bc01 is the ordering of the output of the cudnn conv op.
Currently, pylearn2.linear.conv2d.Conv2D uses b01c as default (even if the Theano conv2d Op it uses outputs bc01), and pylearn2.linear.conv2d_c01b uses c01b (which is also the output of the op).
Using the minimal amount of dimshuffles makes sense to me at the moment, so the native format of the Op would be OK for me.

fvisin replied:
Shuffle should be performed in place by numpy. Actually if I am not wrong, it should not even modify the array in memory, but directly index it in the right way when you access the data. 
Anyway, if you still think it is better to default to bc01 I'm ok with that.

lamblin replied:
I don't think it matters much, as the output_axes are currently always provided explicitly.
It is true that dimshuffles are performed inplace with almost no overhead, and chained dimshuffles would even be fused by Theano. However, there is a slightly higher risk that a "gpu_contiguous" gets inserted between dimshuffles, which would force a copy.
I would keep bc01, but I would not bother making the same change in conv2d.Conv2D.

ballasn replied:
cudnn does call gpu_contiguous, so I think it is important to keep the format bc01 to avoid unecessary copy

fvisin replied:
Ok. Thanks for pointing that out.

fvisin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,195 @@
+            TODO
+        """
+        return np.transpose(self._filters.get_value(borrow=borrow),
+                            (0, 2, 3, 1))
```
This seems to assume filters is provided with a particular axes order. Can you confirm?
In this case I guess it would be better to generalize to whatever input axes where provided.

lamblin replied:
There is no mechanism to provide or use axes for filters, so they need to be in the proper shape.
I don't think we want to introduce such a mechanism.

fvisin replied:
Ok, I forgot the axes order is specified in the docstring so it's ok as is.

fvisin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,195 @@
+        batch_size : TODO
+            TODO
+        """
+        self._img_shape = tuple([batch_size] + list(self._img_shape[1:]))
```
Why is `set_batch_size` modifying `self._img_shape`? And also, why is it not modifying `self._batch_size`?

lamblin replied:
`self._img_shape` contains the shape for the whole input tensor (bc01).
There is no `self._batch_size`.

fvisin commented:
@nouiz I rebased on top of current master, thank you for notifying me that the other PR was merged.

lamblin commented:
The last thing missing for a merge is a unit test.

JesseLivezey commented:
I'm curious, since theano could already insert cudnn convolutions into the graph, is the main purpose of this PR to expose the additional functionality of the cudnn op compared to the conv2d op?

fvisin commented:
The reason for this PR is that Theano, at the moment at least, uses conv3D instead of cudnn if you convolve with stride greater than 1, which is highly inefficient.

lamblin commented:
Theano currently fails to insert cudnn convolution in once case: for the gradient of convolutions with subsampling. In that case, the gradient is computed by CPU-only conv3D.

fvisin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,204 @@
+        img = gpu_contiguous(x)
+        kerns = gpu_contiguous(self._filters)
+        desc = self._desc(img.shape, kerns.shape)
+        rval = self._conv_op(img, kerns, desc)
```
This is not working as it is an implicit call to https://github.com/Theano/Theano/blob/fb4d5b4d50be1cda5ee4c49e9d4b88f0f2f817f3/theano/sandbox/cuda/dnn.py#L419 which needs another parameter (`output`). Can you suggest me a proper way to set `output`?

lamblin replied:
If I remember correctly, "output" is a tensor into which the convolution will be written or accumulated.
You can call `gpu_alloc_empty` with the output shape to create it.

fvisin commented:
Ok, it should be finally done! Let's hope that Travis doesn't complain.
I recreated the same test as for conv2D, excluded those that didn't apply. Some of them are slightly different, but in the overall they should be pretty much the same.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,151 @@
+skip_if_no_gpu()
+
+if cuda.cuda_available is False:
+    raise SkipTest('Optional package cuda disabled')
```
Instead of these last 4 lines, it would be better to call `skip_if_no_gpu()` inside `TestCudnn.setUp`.
This should not interfere with the script checking the docstring format.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,141 @@
+    def test_lmul(self):
+        """
+        Use SciPy's ndimage to check whether the convolution worked
+        correctly
```
Can you update the docstring to say that we compare against conv2d, not scipy?

fvisin replied:
Sure!

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,141 @@
+        )
+        filters = sharedX(filters_values)
+        image = np.random.rand(1, 3, 3, 3).astype(theano.config.floatX)
+        conv2d = Conv2D(filters, 1, input_space)
```
We should test Cudnn2D here instead, right?

fvisin replied:
Ops....sorry I didn't finish my job I guess!

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,141 @@
+        input space are all what we expect
+        """
+        output_space = Conv2DSpace((2, 2), 1)
+        conv2d = make_random_conv2D(1, self.input_space, output_space,
```
Idem here, this is testing the other conv2d.

fvisin replied:
Even worse. Sorry again.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -22,11 +22,21 @@
-from pylearn2.linear import conv2d
+# Try to import the fast cudnn library, else fallback to conv2d
+cudnn2d = None
+if config.device == 'gpu':
```
This will not work when users specify `'gpu0'` or `'gpu1'`, for instance.
A better test would be checking for `theano.sandbox.cuda.cuda_enabled` and `dnn_available`.

fvisin replied:
Ok.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -2988,15 +2994,26 @@ def initialize_transformer(self, rng):
+                    kernel_shape=self.kernel_shape,
+                    subsample=self.kernel_stride,
+                    border_mode=self.border_mode,
+                    rng=rng)
```
If you detect this automatically, maybe it would be better to simply import `cudnn2d` as `conv2d`, and avoid that duplication here.
This would also ensure that `make_sparse_....` would use the cudnn version as well..

fvisin replied:
Ok

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -4455,11 +4473,6 @@ def set_mlp(self, mlp):
-    def get_weights(self):
-
-        return self.raw_layer.get_weights()
-
```
Why did you remove that?

fvisin replied:
Because it's already defined here: https://github.com/fvisin/pylearn2/blob/55939adf102ecbeb3ac239be86e7960e36b25ad0/pylearn2/models/mlp.py#L756-L762

lamblin replied:
But that's not the same class, and not the same implementation

fvisin replied:
Sorry I meant this line: https://github.com/fvisin/pylearn2/blob/55939adf102ecbeb3ac239be86e7960e36b25ad0/pylearn2/models/mlp.py#L4426-L4428

lamblin replied:
Right, OK then.

lamblin commented on pylearn2/models/mlp.py:
```diff
@@ -2961,14 +2967,14 @@ def __init__(self,
 
         assert nonlinearity is not None
-
+        super(ConvElemwise, self).__init__()
```
Why move the call to `super` here?

fvisin replied:
If I am not wrong, it was like that in the code I was porting.
It seems reasonable to check the consistency of the parameters before calling the inherited constructor though, isn't it?

lamblin replied:
Right.

fvisin commented:
Ok, I am going to push the modification you requested. Let me know if I skip the test in the right way please.

lamblin commented:
The way the tests are skipped look correct, I'll test on a machine without cudnn and let you know.
Regarding travis, there are still the following failures related to documentation:
```
======================================================================
FAIL: Test if pep8 is respected.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/lisa-lab/pylearn2/pylearn2/devtools/tests/test_format.py", line 629, in test_format_pep8
    raise AssertionError("PEP8 Format not respected")
AssertionError: PEP8 Format not respected
-------------------- >> begin captured stdout << ---------------------
/home/travis/build/lisa-lab/pylearn2/pylearn2/linear/cudnn2d.py:259:1: W391 blank line at end of file
--------------------- >> end captured stdout << ----------------------
======================================================================
FAIL: Test if docstrings are well formatted.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/lisa-lab/pylearn2/pylearn2/devtools/tests/test_format.py", line 664, in test_format_docstrings
    verify_format_docstrings()
  File "/home/travis/build/lisa-lab/pylearn2/pylearn2/devtools/tests/test_format.py", line 694, in verify_format_docstrings
    raise AssertionError("Docstring format not respected:\n%s" % msg)
AssertionError: Docstring format not respected:
/home/travis/build/lisa-lab/pylearn2/pylearn2/linear/tests/test_cudnn.py:
TestCudnn class has no Parameters section
/home/travis/build/lisa-lab/pylearn2/pylearn2/linear/cudnn2d.py:
make_sparse_random_conv2D:No Parameters section
```

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,146 @@
+        skip_if_no_gpu()
+        if not dnn_available():
+            SkipTest('Skipping tests cause cudnn is not available')
+        self.image = np.random.rand(1, 1, 3, 3).astype(theano.config.floatX)
```
So I tried to run the tests, this fails when floatX is float64 (which is the default).
A solution would be to force floatX to be float32, as is done in https://github.com/fvisin/pylearn2/blob/mlp_cudnn/pylearn2/linear/tests/test_conv2d_c01b.py#L25, and restore that in `tearDown`. I don't think you have to worry about the call to `use` or `unuse`, though.

fvisin replied:
Ok. Everything should be ok now.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,146 @@
+        """
+        skip_if_no_gpu()
+        if not dnn_available():
+            SkipTest('Skipping tests cause cudnn is not available')
```
This should be `raise SkipTest(...)`, actually.

fvisin replied:
Oh, right.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,151 @@
+        Set up a test image and filter to re-use
+        """
+        self.orig_floatX = theano.config.floatX
+        theano.config.floatX = 'float32'
```
This needs to be done after the "SkipTest" exceptions can be raised, otherwise `tearDown` is not called, and the value is not restored.

fvisin replied:
Right, my bad.

lamblin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,259 @@
+        filters_shape=W.get_value(borrow=True).shape,
+        message=message
+    )
+
```
Pep8 says: `W391 blank line at end of file`

fvisin replied:
Sorry, I am messing with my vim settings and it's not behaving as I expect. I will fix it.

lamblin replied:
OK.

lamblin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,259 @@
+    .. todo::
+
+        WRITEME properly
+
```
Missing "Parameters" section.

fvisin replied:
I added it, but I am not sure on the meaning of the parameters so I left them empty.

lamblin commented on pylearn2/linear/cudnn2d.py:
```diff
@@ -0,0 +1,269 @@
+    values are sparse
+
+    Parameter
+    ---------
```
The section should be named "Parameters".
Rather than have empty descriptions, can you put "TODO"?

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,151 @@
+class TestCudnn(unittest.TestCase):
+    """
+    Tests for the Cudnn code
+    """
```
It's complaining about a missing "Parameters" section here, I'm not really sure why.
Maybe you can put a Parameters section that just says to refer to unittest.TestCase and check it fixes the error.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,151 @@
+
+        self.cudnn2d = Cudnn2D(self.filters, self.batch_size, self.input_space)
+
+    def tearDown(self):
```
Missing docstring.

fvisin replied:
Yep, fixed that. Apparently the latest pylint installation on ubuntu broke syntastic :/
I am pushing frequently to find out syntax/pep8 errors with Travis, don't bother pointing them out. I will check Travis logs on my own and I'll comment back when I think I am done, so you don't waste your time on stupid errors like this. :)

lamblin replied:
OK.

fvisin commented:
Finally Travis is happy with it! I ran the tests on GPU with cudnn and they all pass, I didn't check with a GPU that doesn't support cudnn though, if you want to try that again.

lamblin commented:
@fvisin Thanks! I'll test on a GPU without cudnn, and merge if it passes.

lamblin commented on pylearn2/linear/tests/test_cudnn.py:
```diff
@@ -0,0 +1,168 @@
+        """
+        skip_if_no_gpu()
+        if not dnn_available():
+            raise self.SkipTest('Skipping tests cause cudnn is not available')
```
It is just `SkipTest`, not `self.SkipTest`.

lamblin commented:
OK, so the last missing thing is to fix the syntax of `SkipTest`, otherwise the tests fail on a machine with a GPU but not cudnn.

fvisin commented:
Fixed! Should be possible to merge now.

Francesco

On Tue, Apr 28, 2015 at 10:21 PM, Pascal Lamblin <notifications@github.com>
wrote:

> OK, so the last missing thing is to fix the syntax of SkipTest, otherwise
> the tests fail on a machine with a GPU but not cudnn.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lisa-lab/pylearn2/pull/1407#issuecomment-97196267>.
>

lamblin commented:
Thanks a lot!

jstaker7 commented:
Thanks for the fix! I had to change this line in mlp.py:
From:
from linear import cudnn2d as conv2d
To:
from pylearn2.linear import cudnn2d as conv2d

Because linear had not been referenced yet.

fvisin commented:
Thank you @jstaker7 for pointing it out. Should be fixed soon by https://github.com/lisa-lab/pylearn2/pull/1507
